plugin/template.go

```diff
@@ -0,0 +1,284 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"go/parser"
+	"go/printer"
+	"go/token"
+	"sort"
+	"text/template"
+
+	"github.com/thriftrw/thriftrw-go/internal/goast"
+	"github.com/thriftrw/thriftrw-go/plugin/api"
+
+	"golang.org/x/tools/go/ast/astutil"
+)
+
+// TemplateOption provides optional arguments to GoFileFromTemplate.
+type TemplateOption struct {
+	apply func(*goFileGenerator)
+}
+
+// TemplateFunc is a TemplateOption that makes a function available in the
+// template.
+//
+// The function may be anything accepted by text/template.
+//
+// 	GoFileFromTemplate(
+// 		filename,
+// 		`package <lower "HELLO">`,
+// 		TemplateFunc("lower", strings.ToLower),
+// 	)
+func TemplateFunc(name string, f interface{}) TemplateOption {
+	return TemplateOption{apply: func(t *goFileGenerator) {
+		t.funcs[name] = f
+	}}
+}
+
+// GoFileImportPath is a TemplateOption that specifies the intended import path
+// for the file being generated.
+//
+// If specified, this changes the behavior of the `formatType` template function
+// to NOT import this package and instead use the types directly since they are
+// available in the same package.
+func GoFileImportPath(path string) TemplateOption {
+	return TemplateOption{apply: func(t *goFileGenerator) {
+		t.importPath = path
+	}}
+}
+
+// goFileGenerator generates a single Go file.
+type goFileGenerator struct {
+	importPath string
+	funcs      map[string]interface{}
+
+	// Names of known globals. All global variables share this namespace.
+	globals map[string]struct{}
+
+	// import path -> empty string or import name for named imports
+	imports map[string]string
+
+	// import path -> package name or import name for named imports
+	importedNames map[string]string
+}
+
+func newGoFileGenerator(opts []TemplateOption) *goFileGenerator {
+	t := goFileGenerator{
+		funcs:         make(map[string]interface{}),
+		globals:       make(map[string]struct{}),
+		imports:       make(map[string]string),
+		importedNames: make(map[string]string),
+	}
+	for _, opt := range opts {
+		opt.apply(&t)
+	}
+	return &t
+}
+
+// Import the given import path and return the imported name for this package.
+func (g *goFileGenerator) Import(path string) string {
+	if name, ok := g.importedNames[path]; ok {
+		return name
+	}
+
+	name := goast.DeterminePackageName(path)
+
+	// Find an import name that does not conflict with any known globals.
```

Comment:
We should also make sure that the package name is not a reserved keyword or type. E.g. this is totally valid:

$GOPATH/src/foo/bool/bool.go:
```go
package bool

func Test() {
  println("test")
}
```

$GOPATH/src/foo/main.go:
```go
package main

import "foo/bool"

func main() {
  bool.Test()
}
```